core/src/main/java/tc/oc/pgm/filters/parse/FilterParser.java

```diff
@@ -407,6 +412,18 @@ private <T extends FeatureDefinition> XMLFeatureReference<T> reference(Node node
     return this.factory.getFeatures().createReference(node, cls);
   }
 
+  // Warning: this should only be used when you're certain those features load before filters.
+  private <T extends FeatureDefinition> T resolve(Node node, Class<T> cls)
+      throws InvalidXMLException {
+    String id = node.getValueNormalize();
+    T val = this.factory.getFeatures().get(id, cls);
+    if (val == null)
+      throw new InvalidXMLException(
+          "Unknown " + SelfIdentifyingFeatureDefinition.makeTypeName(cls) + " ID '" + id + "'",
+          node);
+    return val;
+  }
+
```

Comment:
Any specific reason this is identical in both parsers but not abstracted out to some `XMLReferenceUtils`?

Reply:
The reason is XMLUtils lie in the `util` package, all of the xml references, FeatureDefinition etc exist only in the `core` package, meaning this sadly cannot be put in utils, where it'd best fit.

Could try to put it in some other xml utils file that resides in core but yeah, it was simpler to copy the one method to both parsers at this time